packages/fxa-auth-server/lib/routes/subscriptions/paypal-notifications.ts

```diff
@@ -31,10 +31,38 @@ export class PayPalNotificationHandler extends PayPalHandler {
       });
     }
 
-    // TODO: If the invoice is void/uncollectible and this is Completed/Processed, issue
-    // a refunded.
-
     if (invoice.status == null || !['draft', 'open'].includes(invoice.status)) {
+      if (
+        invoice.status == 'uncollectible' &&
+        ['Completed', 'Processed'].includes(message.payment_status)
+      ) {
+        // we need to refund the user since the invoice was cancelled
+        // but payment was processed
+        const refundResponse = await this.paypalHelper.refundTransaction({
+          idempotencyKey: message.invoice,
+          transactionId: message.txn_id,
+        });
+        if (
+          ['instant', 'delayed'].includes(
+            refundResponse.refundStatus.toLowerCase()
+          )
+        ) {
+          this.stripeHelper.updateInvoiceWithPaypalRefundTransactionId(
+            invoice,
+            refundResponse.refundTransactionId
+          );
+          return;
+        } else {
+          this.log.error('handleMerchPayment', {
+            message: 'PayPal refund transaction unsuccessful',
+            ipnMessage: message,
+            refundResponse,
+          });
+          throw error.internalValidationError('handleMerchPayment', {
+            message: 'PayPal refund transaction unsuccessful',
+          });
+        }
+      }
```

Comment:
This whole block would be a bit easier to parse if it was its own function, `issueRefund` or something.